### PR TITLE
add function for merging migrations into conda build config and add pinning to all migrations

### LINKED ISSUE #912: Close out migrations
It would be nice to send a PR that closes out a migration (removes or moves the migration yaml and updates the pinning) once a condition has been met. We can default the condition to being all PRs issued. But it might be good to have the condition be configurable.

### PR DESCRIPTION
**Pull request by CJ-Wright**
Closes #912

CC: @mariusvniekerk @beckermr

This introduces a function for merging the conda build config and a migration.
The function preserves selectors by not loading the yamls for either.
This causes the function to only implement a subset of the variant algebra.
However, this should cover most use cases that we currently have, including
the case where `zip_keys` is used and is completely represented in the migraiton.
The API is designed to be flexable so that future implementations can cover more
of the variant algebra.

Overall this makes the migration UX better as it requires fewer human made PRs, with a resonable expection that the bot issued PRs will be correct.
To the best of my knowledge this doesn't disrupt any of our existing UX.

### CODE REVIEW

**Comment by CJ-Wright** [comment]
needs test though

**Comment by CJ-Wright** [comment]
The condition is all PRs have been opened. (essentially adding pinning as a child package of all the packages in the migrtion)

**Comment by beckermr** [comment]
Right. We should add a condition in the migration yaml that say X% are opened or merged etc.

**Comment by CJ-Wright** [comment]
I think that would be a great enhancement but I think it is beyond the scope for this PR.

**Comment by beckermr** [comment]
Sure that's fine.

**Comment by CJ-Wright** [comment]
cycle for ci

**Comment by CJ-Wright** [comment]
@beckermr this passes tests now would you like to review again?

@regro/auto-tick

File: conda_forge_tick/migrators/migration_yaml.py [hunk 1]
```diff
@@ -19,6 +21,37 @@
...
+        # fix for bad indentation from bot's PRs
+        if line.startswith('-'):
+            line = '  ' + line
+        migrator_keys[current_key].append(line)
+
+    outbound_cbc = []
+    current_cbc_key = None
+    for line in conda_build_config_yaml.split("\n"):
+        # if we start with a word/underscore we are starting a section
+        if re.match(r'\w', line) or line.startswith('_') or line.startswith('#'):
```
Comment (beckermr) [inline conda_forge_tick/migrators/migration_yaml.py]: Why are we processing comment lines here with the `#`?

File: conda_forge_tick/migrators/migration_yaml.py [hunk 2]
```diff
@@ -19,6 +21,37 @@
...
+            current_cbc_key = line.split()[0]
+            # if we are in a section from the migrator use the migrator data
+            if current_cbc_key in migrator_keys:
+                outbound_cbc.extend(migrator_keys[line.split()[0]])
+            else:
+                outbound_cbc.append(line)
+        # if in a section that we migrated, ignore
+        elif current_cbc_key not in migrator_keys or line.isspace() or not line:
```
Comment (beckermr) [inline conda_forge_tick/migrators/migration_yaml.py]: This comment and if statement are confusing. Can you explain a bit?

File: conda_forge_tick/migrators/migration_yaml.py [hunk 3]
```diff
@@ -71,6 +104,23 @@ def __init__(
...
+            with indir(recipe_dir):
+                os.remove("conda_build_config.yaml")
+                with open("conda_build_config.yaml", "w") as f:
+                    f.write(merged_cbc)
+            # don't need to bump build number once we move to datetime
+            # version numbers for pinning
+            return super().migrate(recipe_dir, attrs)
+
         # in case the render is old
         os.makedirs(os.path.join(recipe_dir, "../.ci_support"), exist_ok=True)
```
Comment (beckermr) [inline conda_forge_tick/migrators/migration_yaml.py]: Let's put all of the code here and below in an else statement to make the intent clearer

File: tests/test_migration_yaml_migration.py [hunk 4]
```diff
@@ -241,3 +242,16 @@ def run_test_migration(
     actual_output = pat.sub("", actual_output)
     output = pat.sub("", output)
     assert actual_output == output
+
+
+with open(os.path.join(YAML_PATH, 'conda_build_config.yaml'), 'r') as fp:
+    cbc = fp.read()
```
Comment (beckermr) [inline tests/test_migration_yaml_migration.py]: This is effectively a global and so should be in all caps.

File: conda_forge_tick/migrators/migration_yaml.py [hunk 5]
```diff
@@ -71,6 +104,23 @@ def __init__(
     def migrate(
         self, recipe_dir: str, attrs: "AttrsTypedDict", **kwargs: Any
     ) -> "MigrationUidTypedDict":
+        # if conda-forge-pinning do something special
```
Comment (beckermr) [inline conda_forge_tick/migrators/migration_yaml.py]: Let's expand this comment to explain what the special thing is here.

File: conda_forge_tick/migrators/migration_yaml.py (hunk 1 above)
Comment (CJ-Wright) [reply conda_forge_tick/migrators/migration_yaml.py]: They often have no spaces before them so they are at the same indentation, if we didn't handle them then they would not be in the resulting yaml.

File: conda_forge_tick/migrators/migration_yaml.py [hunk 6]
```diff
@@ -19,6 +21,37 @@
...
 
 
+def merge_migrator_cbc(migrator_yaml: str, conda_build_config_yaml: str):
+    """Merge a migrator_yaml with the conda_build_config_yaml"""
+    migrator_keys = defaultdict(list)
+    current_key = None
+    for line in migrator_yaml.split("\n"):
+        if not line or line.isspace():
+            continue
+        if re.match(r"\w", line) or line.startswith("_"):
```
Comment (scopatz) [inline conda_forge_tick/migrators/migration_yaml.py]: You should precompile regexes.  Especially when they are in a loop

File: conda_forge_tick/migrators/migration_yaml.py [hunk 7]
```diff
@@ -71,17 +104,35 @@ def __init__(
...
+        if attrs.get("name", "") == "conda-forge-pinning":
+            # read up the conda build config
+            with indir(recipe_dir), open("conda_build_config.yaml") as f:
+                cbc_contents = f.read()
+            merged_cbc = merge_migrator_cbc(self.yaml_contents, cbc_contents)
+            with indir(os.path.join(recipe_dir, "migrations")):
+                os.remove(f"{self.name}.yaml")
+            # replace the conda build config with the merged one
+            with indir(recipe_dir):
+                os.remove("conda_build_config.yaml")
```
Comment (scopatz) [inline conda_forge_tick/migrators/migration_yaml.py]: Why do the remove when the following line truncates the file in "w"-mode anyway

**Comment by CJ-Wright** [comment]
@beckermr @scopatz ready for another round

**Comment by scopatz** [comment]
LGTM

**Pull request merged**